dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileTiny.cpp

```diff
@@ -183,12 +205,23 @@ Block ColumnFileTiny::readBlockForMinorCompaction(const PageReader & page_reader
     }
 }
 
-ColumnTinyFilePtr ColumnFileTiny::writeColumnFile(DMContext & context, const Block & block, size_t offset, size_t limit, WriteBatches & wbs, const BlockPtr & schema, const CachePtr & cache)
+ColumnTinyFilePtr ColumnFileTiny::writeColumnFile(DMContext & context, const Block & block, size_t offset, size_t limit, WriteBatches & wbs, const CachePtr & cache)
 {
     auto page_id = writeColumnFileData(context, block, offset, limit, wbs);
-    auto new_column_file_schema = schema ? schema : std::make_shared<Block>(block.cloneEmpty());
+
+    auto new_digest = hashSchema(block);
+    auto schema = context.db_context.getSharedBlockSchemas()->find(new_digest);
+
+    std::shared_ptr<ColumnFileInMemory> new_column_file;
+    // Create a new column file.
+    if (schema == nullptr)
+    {
+        schema = std::make_shared<ColumnFileSchema>(block.cloneEmpty());
+        context.db_context.getSharedBlockSchemas()->insert(new_digest, schema);
+    }
```

Comment:
```suggestion
    auto schema_digest = hashSchema(block);
    auto schema = context.db_context.getSharedBlockSchemas()->find(schema_digest);
    // Create a new column file.
    if (schema == nullptr)
    {
        schema = std::make_shared<ColumnFileSchema>(block.cloneEmpty());
        context.db_context.getSharedBlockSchemas()->insert(schema_digest, schema);
    }
```

Reply:
Not sure but this pattern seems to be common, so that there could be a function doing this:

```c++
auto schema = context.db_context.getSharedBlockSchemas().getSchemaFromBlock(block);
```